File: backend/app/api/v1/endpoints/clientes.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Optional, List
from uuid import UUID
from datetime import date

from app.core.deps import get_db, get_usuario_actual
from app.models.usuario import Usuario
from app.models.cliente import TipoCliente
from app.schemas.cliente import (
    ClienteCreate, ClienteUpdate, ClienteResponse, ClienteListItem,
    EstadoCuentaCliente, MovimientoCuentaCorriente, RegistroPagoCreate,
    PagoResponse, UsuarioClienteCreate, UsuarioClienteResponse,
    ProyectoClienteResumen, ClienteConProyectos
)
from app.services import cliente_service

router = APIRouter()


# ============ CRUD CLIENTES ============

@router.get("", response_model=List[ClienteListItem])
def listar_clientes(
    tipo: Optional[TipoCliente] = None,
    activo: Optional[bool] = True,
    con_deuda: Optional[bool] = None,
    buscar: Optional[str] = None,
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Lista todos los clientes con filtros."""
    clientes = cliente_service.get_clientes(
        db, tipo=tipo, activo=activo, con_deuda=con_deuda,
        buscar=buscar, skip=skip, limit=limit
    )

    return [
        ClienteListItem(
            id=c.id,
            codigo=c.codigo,
            razon_social=c.razon_social,
            nombre_fantasia=c.nombre_fantasia,
            tipo=c.tipo,
            cuit=c.cuit,
            telefono=c.telefono,
            email=c.email,
            ciudad=c.ciudad,
            saldo_cuenta_corriente=float(c.saldo_cuenta_corriente or 0),
            activo=c.activo
        )
        for c in clientes
    ]


@router.get("/lista", response_model=List[ClienteListItem])
def lista_clientes_simple(
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Lista simplificada de clientes para selectores."""
    return cliente_service.get_clientes_lista(db)


@router.get("/tipos")
def obtener_tipos_cliente():
    """Obtiene los tipos de cliente disponibles."""
    return [{"value": t.value, "label": t.value.replace("_", " ").title()} for t in TipoCliente]


@router.get("/condiciones-iva")
def obtener_condiciones_iva():
    """Obtiene las condiciones de IVA disponibles."""
    from app.models.cliente import CondicionIVA
    labels = {
        "responsable_inscripto": "Responsable Inscripto",
        "monotributo": "Monotributo",
        "exento": "Exento",
        "consumidor_final": "Consumidor Final",
        "no_responsable": "No Responsable"
    }
    return [{"value": c.value, "label": labels.get(c.value, c.value)} for c in CondicionIVA]


@router.get("/estadisticas")
def obtener_estadisticas(
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Obtiene estadísticas generales de clientes."""
    return cliente_service.get_estadisticas_clientes(db)


@router.get("/{cliente_id}", response_model=ClienteResponse)
def obtener_cliente(
    cliente_id: UUID,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Obtiene el detalle de un cliente."""
    cliente = cliente_service.get_cliente(db, cliente_id)
    if not cliente:
        raise HTTPException(status_code=404, detail="Cliente no encontrado")

    return cliente_service.get_cliente_response(db, cliente)


@router.get("/{cliente_id}/completo", response_model=ClienteConProyectos)
def obtener_cliente_completo(
    cliente_id: UUID,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Obtiene el detalle completo de un cliente con proyectos."""
    cliente = cliente_service.get_cliente_con_proyectos(db, cliente_id)
    if not cliente:
        raise HTTPException(status_code=404, detail="Cliente no encontrado")

    return cliente


@router.post("", response_model=ClienteResponse)
def crear_cliente(
    cliente: ClienteCreate,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Crea un nuevo cliente."""
    try:
        db_cliente = cliente_service.create_cliente(db, cliente)
        return cliente_service.get_cliente_response(db, db_cliente)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.put("/{cliente_id}", response_model=ClienteResponse)
def actualizar_cliente(
    cliente_id: UUID,
    cliente: ClienteUpdate,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Actualiza un cliente existente."""
    try:
        db_cliente = cliente_service.update_cliente(db, cliente_id, cliente)
        if not db_cliente:
            raise HTTPException(status_code=404, detail="Cliente no encontrado")

        return cliente_service.get_cliente_response(db, db_cliente)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/{cliente_id}")
def eliminar_cliente(
    cliente_id: UUID,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Desactiva un cliente (soft delete)."""
    try:
        success = cliente_service.delete_cliente(db, cliente_id)
        if not success:
            raise HTTPException(status_code=404, detail="Cliente no encontrado")

        return {"message": "Cliente eliminado correctamente"}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


# ============ CUENTA CORRIENTE ============

@router.get("/{cliente_id}/cuenta-corriente", response_model=EstadoCuentaCliente)
def obtener_estado_cuenta(
    cliente_id: UUID,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Obtiene el estado de cuenta de un cliente."""
    estado = cliente_service.get_estado_cuenta(db, cliente_id)
    if not estado:
        raise HTTPException(status_code=404, detail="Cliente no encontrado")

    return estado


@router.get("/{cliente_id}/movimientos", response_model=List[MovimientoCuentaCorriente])
def obtener_movimientos(
    cliente_id: UUID,
    fecha_desde: Optional[date] = None,
    fecha_hasta: Optional[date] = None,
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Obtiene los movimientos de cuenta corriente de un cliente."""
    return cliente_service.get_movimientos_cuenta(
        db, cliente_id,
        fecha_desde=fecha_desde,
        fecha_hasta=fecha_hasta,
        skip=skip, limit=limit
    )


@router.post("/{cliente_id}/pagos", response_model=PagoResponse)
def registrar_pago(
    cliente_id: UUID,
    pago: RegistroPagoCreate,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Registra un pago de un cliente."""
    try:
        return cliente_service.registrar_pago(db, cliente_id, pago, current_user.id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


# ============ USUARIO CLIENTE ============

@router.get("/{cliente_id}/usuario", response_model=Optional[UsuarioClienteResponse])
def obtener_usuario_cliente(
    cliente_id: UUID,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Obtiene el usuario asociado a un cliente."""
    return cliente_service.get_usuario_cliente(db, cliente_id)


@router.post("/{cliente_id}/usuario", response_model=UsuarioClienteResponse)
def crear_usuario_cliente(
    cliente_id: UUID,
    datos: UsuarioClienteCreate,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Crea un usuario de acceso al portal para el cliente."""
    try:
        return cliente_service.crear_usuario_cliente(db, cliente_id, datos)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{cliente_id}/usuario/reset-password")
def reset_password_usuario(
    cliente_id: UUID,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Resetea la contraseña del usuario cliente."""
    try:
        nuevo_password = cliente_service.reset_password_usuario_cliente(db, cliente_id)
        return {"password": nuevo_password}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


# ============ PROYECTOS ============

@router.get("/{cliente_id}/proyectos", response_model=List[ProyectoClienteResumen])
def obtener_proyectos_cliente(
    cliente_id: UUID,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_usuario_actual)
):
    """Obtiene los proyectos de un cliente."""
    return cliente_service.get_proyectos_cliente(db, cliente_id)

File: frontend/src/services/clientes.ts
import api from './api'

// ============ TYPES ============

export type TipoCliente = 'particular' | 'empresa' | 'gobierno' | 'constructora' | 'consorcio' | 'otro'
export type CondicionIVA = 'responsable_inscripto' | 'monotributo' | 'exento' | 'consumidor_final' | 'no_responsable'

export interface Cliente {
  id: string
  codigo: string
  tipo: TipoCliente
  razon_social: string
  nombre_fantasia?: string
  cuit?: string
  condicion_iva: CondicionIVA
  email?: string
  telefono?: string
  celular?: string
  contacto_nombre?: string
  contacto_cargo?: string
  direccion?: string
  ciudad?: string
  provincia?: string
  codigo_postal?: string
  descuento_general: number
  limite_credito?: number
  dias_credito: number
  enviar_notificaciones: boolean
  notas?: string
  notas_internas?: string
  saldo_cuenta_corriente: number
  fecha_alta?: string
  fecha_ultimo_proyecto?: string
  usuario_id?: string
  activo: boolean
  nombre_display: string
  tiene_deuda: boolean
  cantidad_proyectos: number
  usuario_email?: string
}

export interface ClienteListItem {
  id: string
  codigo: string
  razon_social: string
  nombre_fantasia?: string
  tipo: TipoCliente
  cuit?: string
  telefono?: string
  email?: string
  ciudad?: string
  saldo_cuenta_corriente: number
  activo: boolean
}

export interface ClienteCreate {
  tipo?: TipoCliente
  razon_social: string
  nombre_fantasia?: string
  cuit?: string
  condicion_iva?: CondicionIVA
  email?: string
  telefono?: string
  celular?: string
  contacto_nombre?: string
  contacto_cargo?: string
  direccion?: string
  ciudad?: string
  provincia?: string
  codigo_postal?: string
  descuento_general?: number
  limite_credito?: number
  dias_credito?: number
  enviar_notificaciones?: boolean
  notas?: string
  notas_internas?: string
}

export interface ClienteUpdate {
  tipo?: TipoCliente
  razon_social?: string
  nombre_fantasia?: string
  cuit?: string
  condicion_iva?: CondicionIVA
  email?: string
  telefono?: string
  celular?: string
  contacto_nombre?: string
  contacto_cargo?: string
  direccion?: string
  ciudad?: string
  provincia?: string
  codigo_postal?: string
  descuento_general?: number
  limite_credito?: number
  dias_credito?: number
  enviar_notificaciones?: boolean
  notas?: string
  notas_internas?: string
}

export interface EstadoCuentaCliente {
  cliente_id: string
  cliente_nombre: string
  saldo_actual: number
  limite_credito?: number
  credito_disponible?: number
  total_facturado_mes: number
  total_pagado_mes: number
  proyectos_activos: number
  dias_factura_mas_antigua?: number
}

export interface MovimientoCuentaCorriente {
  id: string
  tipo: 'cargo' | 'pago' | 'ajuste'
  concepto: string
  monto: number
  fecha: string
  saldo_anterior: number
  saldo_posterior: number
  referencia?: string
  proyecto_nombre?: string
}

export interface RegistroPagoCreate {
  monto: number
  fecha: string
  metodo_pago?: string
  referencia?: string
  notas?: string
  proyecto_id?: string
}

export interface PagoResponse {
  id: string
  cliente_id: string
  monto: number
  fecha: string
  metodo_pago: string
  referencia?: string
  numero_recibo: string
  saldo_anterior: number
  saldo_posterior: number
}

export interface UsuarioClienteCreate {
  email: string
  nombre: string
  apellido?: string
  telefono?: string
  password?: string
}

export interface UsuarioClienteResponse {
  id: string
  email: string
  nombre: string
  apellido?: string
  activo: boolean
  password_visible?: string
}

export interface ProyectoClienteResumen {
  id: string
  nombre: string
  estado: string
  porcentaje_avance: number
  fecha_inicio?: string
  fecha_fin_estimada?: string
  monto_contratado?: number
}

export interface ClienteConProyectos extends Cliente {
  proyectos: ProyectoClienteResumen[]
}

export interface EstadisticasClientes {
  total_clientes: number
  clientes_con_deuda: number
  total_deuda: number
  por_tipo: { tipo: string; cantidad: number }[]
}

// ============ API FUNCTIONS ============

// CRUD Clientes
export const getClientes = async (params?: {
  tipo?: TipoCliente
  activo?: boolean
  con_deuda?: boolean
  buscar?: string
  skip?: number
  limit?: number
}): Promise<ClienteListItem[]> => {
  const response = await api.get('/clientes', { params })
  return response.data
}

export const getClientesLista = async (): Promise<ClienteListItem[]> => {
  const response = await api.get('/clientes/lista')
  return response.data
}

export const getCliente = async (id: string): Promise<Cliente> => {
  const response = await api.get(`/clientes/${id}`)
  return response.data
}

export const getClienteCompleto = async (id: string): Promise<ClienteConProyectos> => {
  const response = await api.get(`/clientes/${id}/completo`)
  return response.data
}

export const createCliente = async (data: ClienteCreate): Promise<Cliente> => {
  const response = await api.post('/clientes', data)
  return response.data
}

export const updateCliente = async (id: string, data: ClienteUpdate): Promise<Cliente> => {
  const response = await api.put(`/clientes/${id}`, data)
  return response.data
}

export const deleteCliente = async (id: string): Promise<void> => {
  await api.delete(`/clientes/${id}`)
}

// Tipos y condiciones
export const getTiposCliente = async (): Promise<{ value: string; label: string }[]> => {
  const response = await api.get('/clientes/tipos')
  return response.data
}

export const getCondicionesIVA = async (): Promise<{ value: string; label: string }[]> => {
  const response = await api.get('/clientes/condiciones-iva')
  return response.data
}

// Estadísticas
export const getEstadisticasClientes = async (): Promise<EstadisticasClientes> => {
  const response = await api.get('/clientes/estadisticas')
  return response.data
}

// Cuenta Corriente
export const getEstadoCuenta = async (clienteId: string): Promise<EstadoCuentaCliente> => {
  const response = await api.get(`/clientes/${clienteId}/cuenta-corriente`)
  return response.data
}

export const getMovimientosCuenta = async (
  clienteId: string,
  params?: {
    fecha_desde?: string
    fecha_hasta?: string
    skip?: number
    limit?: number
  }
): Promise<MovimientoCuentaCorriente[]> => {
  const response = await api.get(`/clientes/${clienteId}/movimientos`, { params })
  return response.data
}

export const registrarPago = async (clienteId: string, data: RegistroPagoCreate): Promise<PagoResponse> => {
  const response = await api.post(`/clientes/${clienteId}/pagos`, data)
  return response.data
}

// Usuario Cliente
export const getUsuarioCliente = async (clienteId: string): Promise<UsuarioClienteResponse | null> => {
  const response = await api.get(`/clientes/${clienteId}/usuario`)
  return response.data
}

export const crearUsuarioCliente = async (
  clienteId: string,
  data: UsuarioClienteCreate
): Promise<UsuarioClienteResponse> => {
  const response = await api.post(`/clientes/${clienteId}/usuario`, data)
  return response.data
}

export const resetPasswordUsuario = async (clienteId: string): Promise<{ password: string }> => {
  const response = await api.post(`/clientes/${clienteId}/usuario/reset-password`)
  return response.data
}

// Proyectos del cliente
export const getProyectosCliente = async (clienteId: string): Promise<ProyectoClienteResumen[]> => {
  const response = await api.get(`/clientes/${clienteId}/proyectos`)
  return response.data
}

// ============ HELPERS ============

export const getTipoClienteLabel = (tipo: TipoCliente): string => {
  const labels: Record<TipoCliente, string> = {
    particular: 'Particular',
    empresa: 'Empresa',
    gobierno: 'Gobierno',
    constructora: 'Constructora',
    consorcio: 'Consorcio',
    otro: 'Otro',
  }
  return labels[tipo] || tipo
}

export const getCondicionIVALabel = (condicion: CondicionIVA): string => {
  const labels: Record<CondicionIVA, string> = {
    responsable_inscripto: 'Responsable Inscripto',
    monotributo: 'Monotributo',
    exento: 'Exento',
    consumidor_final: 'Consumidor Final',
    no_responsable: 'No Responsable',
  }
  return labels[condicion] || condicion
}

export const getTipoClienteColor = (tipo: TipoCliente): string => {
  const colors: Record<TipoCliente, string> = {
    particular: 'bg-blue-100 text-blue-800',
    empresa: 'bg-purple-100 text-purple-800',
    gobierno: 'bg-amber-100 text-amber-800',
    constructora: 'bg-orange-100 text-orange-800',
    consorcio: 'bg-teal-100 text-teal-800',
    otro: 'bg-gray-100 text-gray-800',
  }
  return colors[tipo] || 'bg-gray-100 text-gray-800'
}

export const formatCUIT = (cuit?: string): string => {
  if (!cuit) return '-'
  // Formato: XX-XXXXXXXX-X
  const clean = cuit.replace(/\D/g, '')
  if (clean.length !== 11) return cuit
  return `${clean.slice(0, 2)}-${clean.slice(2, 10)}-${clean.slice(10)}`
}
